Fix for importaddress returning incorrect amounts

francisjyap opened the pull request:
Fix for importaddress returning incorrect amounts

This is for Issue #193

aguycalled commented:
Maybe the issue is not clear enough, but I think he refers to an issue when importing cold staking addresses.

francisjyap commented:
@aguycalled Updated the test case to include cold staking addresses.

aguycalled commented:
just tested it, importing the spending or staking address in the second node does not show the balance as if it was an address owned by the wallet.

francisjyap commented:
@aguycalled oh, i assumed that the balance from addresses imported were not to be shown as if it was owned by the wallet, but only show transactions on the watched address. In this case, should commands like getbalance be updated to include balances from watched addresses?

aguycalled commented:
you are right, only the transactions are watched, but those do not appear for spending or staking addresses.

mxaddict commented:
> you are right, only the transactions are watched, but those do not appear for spending or staking addresses.

So does that mean:

```
ZZZ = normal address # Has balance
XXX = staking # No balance
YYY = spending # No balance
[YYY,XXX] = cold address # Has balance

Import ZZZ = works (You see transactions + correct amount)
Import XXX = (does this work?)
Import YYY = (does this work?)
Import [YYY,XXX] = works (You see transactions + correct amount)
```

aguycalled commented:
> > you are right, only the transactions are watched, but those do not appear for spending or staking addresses.
> 
> So does that mean:
> 
> ```
> ZZZ = normal address # Has balance
> XXX = staking # No balance
> YYY = spending # No balance
> [YYY,XXX] = cold address # Has balance
> 
> Import ZZZ = works (You see transactions + correct amount)
> Import XXX = (does this work?)
> Import YYY = (does this work?)
> Import [YYY,XXX] = works (You see transactions + correct amount)
> ```

yes

mxaddict commented:
@francisjyap I think you just need to add those scenarios to the functional test, then fix the issue :+1: if it still exists.

@aguycalled  so the balance is not supposed to update, only the transactions list correct?

aguycalled commented:
watchonly address balance is only shown in qt iirc

francisjyap commented:
@mxaddict @aguycalled I updated the test cases per your suggestions and the transactions and balance from the imported addresses are reflected correctly.

I added a Watched label on the UI showing the total balance from watched addresses. The Total label is also updated to reflect the total balance plus the total watched balance.

mxaddict commented:
utACK

On Wed, May 15, 2019, 19:45 Francis Yap <notifications@github.com> wrote:

> @mxaddict <https://github.com/mxaddict> @aguycalled
> <https://github.com/aguycalled> I updated the test cases per your
> suggestions and the transactions and balance from the imported addresses
> are reflected correctly.
>
> I added a Watched label on the UI showing the total balance from watched
> addresses. The Total label is also updated to reflect the total balance
> plus the total watched balance.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/479?email_source=notifications&email_token=AAIDAKJWLCCBY4BDLZY7423PVPZU3A5CNFSM4HMBF7Y2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVOMXMQ#issuecomment-492620722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIDAKKSVHRJHSYMDMT7LM3PVPZU3ANCNFSM4HMBF7YQ>
> .
>

mxaddict commented on qa/rpc-tests/importaddress.py:
```diff
@@ -0,0 +1,71 @@
+
+
+if __name__ == '__main__':
+    GetStakeReport().main()
```
Needs to match the name of the file

aguycalled commented on qa/rpc-tests/importaddress.py:
```diff
@@ -0,0 +1,71 @@
+        self.nodes[2].importaddress(address)
+        self.nodes[2].importaddress(spending_address)
+        self.nodes[2].importaddress(staking_address)
+        self.nodes[2].importaddress(coldstaking_address)
```
if you import both the spending/staking and the cold staking address, you can't discern if the txs are seen because the wallet is watching the cold staking address (which was already working) or because it is watching the spending/staking addresses (which were not working before). I'd suggest a setup with one node for each case.

mxaddict commented:
@francisjyap any updates on this issue?

francisjyap commented:
@aguycalled @mxaddict So i found out what was causing the cold staking transactions from appearing when importing a staking/spending address. Cold staking txs would only compare against addresses that are owned by the wallet and not against watched addresses. I also updated the test case.

mxaddict commented:
utACK

aguycalled commented:
utACK

mxaddict commented:
Full test suite + new test cases for this scenario also pass

aguycalled commented:
Tested OSX

aguycalled commented:
paid 5ab73dece539ab5a29b196a4af43e1834c9a5e35da1c977b92d71a005fb4feed
